src/algebra/cubic_solution.lean

```diff
@@ -0,0 +1,570 @@
+/-
+Copyright (c) 2021 Nicholas Dyson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Nicholas Dyson.
+-/
+
+import algebra.invertible
+import tactic.linarith
+import tactic.field_simp
+import order.filter.at_top_bot
+
+/-!
+# Cubic solutions
+
+In this file we define and prove the three solutions to a general cubic in a general
+field that is not of characteristic two or three, and uniqueness of these solutions.
+
+## Main results
+
+`cubic_solution` : The solution to a cubic
+`cubic_solution_unique` : The three solutions generated by using the three cube roots in the
+                          formula are the only solutions.
+-/
+
+
+section fields
+variables {F : Type*}
+
+/--
+((-1 + sqrt(-3))/2)^3 = 1
+-/
+def cube_root_unity_a [field F] (root_three_i : F) : F :=
+  (-1 + root_three_i)/2
+
+/--
+((-1 - sqrt(-3))/2)^3 = 1
+-/
+def cube_root_unity_b [field F] (root_three_i : F) : F :=
+  (-1 - root_three_i)/2
+
+lemma multiply_out_cubed [field F] (x y : F) : (x + y) * (x + y) * (x + y) =
+        (x * x * x) + (3 * x * x * y) + (3 * x * y * y) + (y * y * y) :=
+begin
+  ring,
+end
```

Comment:
This lemma probably eventually belongs next to `add_pow_two` in `algebra/group_power/basic`, with the name `add_pow_three`, but that's probably best to move later once everything else is reviewed.